Change payment processing for subscriptions

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
Saved payment methods are now correctly displayed - previously they were hidden based on whether or not the 'Save payment method' checkbox was visible, by running the `wc_stripe_display_save_payment_method_checkbox` filter.

The checkbox display was ultimately decided by the Stripe extension, which set the checkbox to not display if a subscription product was in the cart. The logic was because the payment method should always be saved when buying a subscription, so displaying the checkbox was unnecessary.

To fix this I have changed the naming of some props used when registering payment methods:
- Previously, we used `savePaymentInfo` to inform the UI whether we should show saved cards **and** the checkbox. 
- I have now changed this to `showSavedCards` and `showSaveOption` to give more granular control over the UI to the payment methods. In our Stripe integration, we set `showSavedCards` based on the value of the `saved_cards` option in the Stripe extension, and we set `showSaveOption` based on the result of the `wc_stripe_display_save_payment_method_checkbox` filter.
- Change `get_allow_saved_cards` method in the `Stripe` class to be named `get_show_saved_cards` and return whether the `saved_cards` option is enabled, without running the checkbox filter mentioned above.
- Add a method to the Stripe class `get_show_save_option` which will return the result of `get_show_saved_cards` filtered through `wc_stripe_display_save_payment_method_checkbox`
- Change the `PaymentMethodTab` component to accept a boolean `showSaveOption` prop. It is this prop that determines if the checkbox should be displayed or not.
- Update the payment method registration code to reflect the new names of these props.
- Update the payment method config and validation to reflect the new names of these props. I also included a shim to set `supports.showSavedCards` to the same value as `supports.savePaymentInfo` for backward compatibility. If any payment methods are registering and still using `savePaymentInfo` then a deprecation notice will be emitted using `@wordpress/deprecated`.
- Updated test files to reflect the new names of these props.

I have also modified the Checkout API so that it will fire a new action `__experimental_woocommerce_blocks_checkout_update_order_meta` which is handled in Subscriptions - this is required as extensions should be given the chance to update the metadata for the orders.

<!-- Reference any related issues or PRs here -->
Fixes #3616
Fixes #3620 
Fixes #3271 
Related PR in subs: 3963-gh-woocommerce/woocommerce-subscriptions

### How to test the changes in this Pull Request

1. Check out this branch and `add/actions-hooks-for-update-meta-from-blocks` on `woocommerce-subscriptions`
1. Enable a payment gateway that supports subs and blocks, for example Stripe 💳
2. Add a non-subscription product to your cart. 
3. Go to the checkout block and when using the Stripe payment method verify that the `Save payment method` checkbox shows.
4. Empty your cart.
4. Create a subscription product that has multiple variations so that you may up/down/crossgrade between them. 
3. Add a subscription for this product to your cart and check out using the checkout block, and ensure any saved cards show as options.
4. Ensure the checkout block asks you for payment, or allows you to use an existing method if you have one saved.
5. Ensure the subscription is created correctly with the correct variation, renewal dates, and price. Also ensure the payment method used to create the subscription was recorded.
6. Go to My account -> Subscription -> and begin the process to up/down/crossgrade the subscription.
7. Go to the checkout block and ensure you are not asked for payment. The payment method should already be saved on the subscription. (If you have prorate upgrades and downgrades turned on you may actually be asked for payment here)
8. Check out, and ensure the checkout completes and you can see the order summary showing your subscription on the client.
9. Visit the subscription in the admin dashboard and ensure it has the correct details for the order you just created.
10. Ensure there is a singular additional order in the Orders list for the *grade we just performed.

opr committed:
Remove savePaymentInfo check when displaying payment methods

This is because the savePaymentInfo is derived from whether the save payment method checkbox shows. This check doesn't make sense to do because it's not a good indicator of whether the payment method is enabled. Subscriptions for example hides the checkbox because it is implied that the method will be saved. We should instead rely on the server-side to only send permitted saved payment methods.

opr committed:
Add safely_get_request_payment_method

This will allow us to try to get the payment method if it was passed in the request, but will default to an empty string if not. This is different to get_request_payment_method because it doesn't throw any errors. We need it to be different because get_request_payment_method is used when the order definitely needs payment (so a normal checkout scenario, vs. a £0 subscription checkout)

opr committed:
Add action to update order meta when checking out

This is needed because some extensions rely on this action to add their information to the metadata of order items.

opr committed:
Remove safely_get_request_payment_method

This is no longer needed.

opr committed:
Remove @since from experimental hook

opr committed:
Add PHPDoc for new update_order_meta hook

opr committed:
Document use of experimental hook

budzanowski commented:
> This is not a reliable check, and I don't think it should be the job of the client to decide which payment methods to display. We should rely on the server-side to send back only valid payment methods. This may need testing to ensure only valid payment methods for subscriptions can be selected.

With the check removed, we will surface the saved payment methods even if they are not allowed. If we would want to remove the check here we should ensure that the payment methods are not sending the saved options otherwise we are exposing them to the user even if they are not working. This is how I understand the problem.

nerrad commented:
> If we would want to remove the check here we should ensure that the payment methods are not sending the saved options otherwise we are exposing them to the user even if they are not working.

I agree with Bartek here. Thomas did you confirm that the server is only surfacing saved payment methods for payment methods that are active? If not, we will need to make sure this is fixed on the server side.

senadir commented on assets/js/base/context/cart-checkout/payment-methods/payment-method-data-context.js:
```diff
@@ -91,13 +91,7 @@ const getCustomerPaymentMethods = ( availablePaymentMethods = {} ) => {
 	const enabledCustomerPaymentMethods = {};
 	paymentMethodKeys.forEach( ( type ) => {
 		const methods = customerPaymentMethods[ type ].filter(
-			( { method: { gateway } } ) => {
-				const isAvailable = gateway in availablePaymentMethods;
-				return (
-					isAvailable &&
-					availablePaymentMethods[ gateway ].supports?.savePaymentInfo
-				);
-			}
```
It's still fuzzy to me what this check does exactly? and why did it had `supports?.savePaymentInfo`?

senadir commented on src/StoreApi/Routes/Checkout.php:
```diff
@@ -375,6 +374,23 @@ private function create_or_update_draft_order() {
 			$order_controller->update_order_from_cart( $this->order );
 		}
 
+		/**
+		 * WooCommerce Blocks Checkout Update Order Meta (experimental).
+		 *
+		 * This hook gives extensions the chance to add or update meta data on the $order.
+		 *
+		 * This is similar to existing core hook woocommerce_checkout_update_order_meta.
+		 * We're using a new action:
+		 * - To keep the interface focused (only pass $order, not passing request data).
+		 * - This also explicitly indicates these orders are from checkout block/StoreAPI.
+		 *
+		 * @see https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3686
+		 * @internal This Hook is experimental and may change or be removed.
+		 *
+		 * @param WC_Order $order Order object.
+		 */
+		do_action( '__experimental_woocommerce_blocks_checkout_update_order_meta', $this->order );
+
```
We did a lot of testing yesterday in our call so I'm confident that this works, I can't comment yet on the payment change until I understand it a bit better but approving this part.

opr committed:
Reinstate the check for allowing saved cards

opr committed:
Add method to Stripe integration to determine if saved_cards is enabled

opr committed:
Add new field to get_payment_method_data

This adds displaySavePaymentMethodCheckbox which will be used to determine if the checkbox to save payment methods should display.

opr committed:
Add displaySavePaymentMethodCheckbox option to client

This will determine whether the "Save payment information" checkbox will be displayed.

opr committed:
Add requiresSaving option to Stripe payment method data

This is informed by the saved_cards option and the result of the wc_stripe_display_save_payment_method_checkbox filter.

opr committed:
Rename displaySavePaymentMethodCheckbox to requiresSaving & fix logic

budzanowski commented on assets/js/blocks-registry/payment-methods/payment-method-config.js:
```diff
@@ -86,5 +87,14 @@ export default class PaymentMethodConfig {
 		}
+		if (
+			config.supports &&
+			typeof config.supports.requiresSaving !== 'undefined' &&
```
If you use optional chaining here you can get rid of the previous line.

opr replied:
Great idea, let's add this as a cooldown task to tidy this up so we don't pollute the PR. Thanks for the tip!

opr replied:
Never mind, I see how small this change is, I thought it was going to be much larger. I've added it to the PR!

nerrad commented on assets/js/base/components/payment-methods/payment-method-tab.js:
```diff
@@ -21,11 +21,13 @@ import PaymentMethodErrorBoundary from './payment-method-error-boundary';
  * @param {boolean} props.allowsSaving Whether that payment method allows saving
  *                                     the data for future purchases.
+ * @param {boolean} props.requiresSaving Whether the payment method should display the option to save
+ * 														   the details entered by the customer.
```
nit, comment block spacing here seems off.

nerrad commented on assets/js/base/components/payment-methods/payment-method-tab.js:
```diff
@@ -36,7 +38,7 @@ const PaymentMethodTab = ( { children, allowsSaving } ) => {
 		<PaymentMethodErrorBoundary isEditor={ isEditor }>
 			{ children }
-			{ customerId > 0 && allowsSaving && (
+			{ customerId > 0 && allowsSaving && ! requiresSaving && (
```
hmm this conditional seems a bit wonky to me. Don't we want something like:
```suggestion
			{ customerId > 0 && ( allowsSaving || requiresSaving ) && (
```

In other words, "If a customer is logged in, and the payment method allows saving or requires saving, then show the checkbox".

nerrad commented on src/Payments/Integrations/Stripe.php:
```diff
@@ -104,12 +105,22 @@ public function get_payment_method_data() {
 	 */
 	private function get_allow_saved_cards() {
-		$saved_cards = isset( $this->settings['saved_cards'] ) ? $this->settings['saved_cards'] : false;
+		return isset( $this->settings['saved_cards'] ) ? 'yes' === $this->settings['saved_cards'] : false;
```
nice little improvement here :)

nerrad commented on src/Payments/Integrations/Stripe.php:
```diff
@@ -104,12 +105,22 @@ public function get_payment_method_data() {
 		// See https://github.com/woocommerce/woocommerce-gateway-stripe/blob/ad19168b63df86176cbe35c3e95203a245687640/includes/class-wc-gateway-stripe.php#L271 and
 		// https://github.com/woocommerce/woocommerce/wiki/Payment-Token-API .
-		return apply_filters( 'wc_stripe_display_save_payment_method_checkbox', filter_var( $saved_cards, FILTER_VALIDATE_BOOLEAN ) );
+		return ! apply_filters( 'wc_stripe_display_save_payment_method_checkbox', filter_var( $saved_cards, FILTER_VALIDATE_BOOLEAN ) );
```
hmm this seems to indicate to me that I might be confused about what `requiresSaving` means. I'm a bit wary of the implicit connection between `requiresSaving` and `allowsSavedCards`. So essentially what we're saying here is that if anything filters this option so it returns false, that means the the saved payment methods ARE required (double negative resulting in that).  I wonder if we should introduce a new filter that allows for explicitly setting this rather than implicitly?

senadir replied:
> I wonder if we should introduce a new filter that allows for explicitly setting this rather than implicitly?

double down on this.

nerrad replied:
Thomas and I chatted about this in our call and reviewed existing behaviour in the shortcode flow. It turns out we think we can improve this by retaining the explicit use of the filtered value and non filtered value and update the name of the properties to make it more clear how things are used. He's in the process of doing that now. My concerns and wariness with the current approach ended up being more an artifact of how the filtered value was returned (switching boolean) and the naming of the properties/values involved.

nerrad reviewed (commented):
Thanks for tackling this Thomas! I have some feedback that might be due to a lack of fully understanding on my part. As noted in one of the inline comments, I have some concern about the implicit assumptions being made by the usage of the filter.

opr committed:
Revert negation on display_save_payment method_checkbox filter & rename

We are going to rename the properties we use to determine whether saved cards are shown, or whether the save payment method checkbox is shown, so that their names are more descriptive of what they are for.

opr committed:
Rename allowSavedCards and requiresSaving in Stripe integration

opr committed:
Rename savePaymentInfo&requiresSaving to showSavedCards & showSaveOption

This is so we can hide the checkbox independently of hiding the saved payment methods.

opr committed:
Show deprecated message if payment methods use savePaymentInfo

This is because we are leaving it in to enable backward compatibility but payment methods registering using this should be informed of the change in case it gets removed.

opr committed:
Update Stripe typedefs and keys of supports object

opr committed:
Show customer payment methods if showSavedCards is true on the method

opr committed:
Make PaymentMethodTab accept showSaveOption prop

This will allow us to show the save checkbox only if the payment method says it should be shown.

opr committed:
Update tests to use new keys in supports when reg'ing payment methods

opr committed:
Add optional chaining when validating payment method config

This makes the code a little tidier :)

opr commented:
Thanks for the feedback everyone, I have updated the testing instructions and PR body to include more information about what changes I have made. Would you mind taking another look and giving it a quick spin?

senadir commented on assets/js/blocks-registry/payment-methods/payment-method-config.js:
```diff
@@ -21,7 +26,11 @@ export default class PaymentMethodConfig {
-			savePaymentInfo: config?.supports?.savePaymentInfo || false,
+			showSavedCards:
+				config?.supports?.showSavedCards ||
+				config?.supports?.savePaymentInfo || // Kept for backward compatibility if methods still pas this when registering.
```
```suggestion
				config?.supports?.savePaymentInfo || // Kept for backward compatibility if methods still pass this when registering.
```

senadir commented on assets/js/blocks-registry/payment-methods/payment-method-config.js:
```diff
@@ -78,12 +87,24 @@ export default class PaymentMethodConfig {
+			);
+		}
+		if ( typeof config.supports?.savePaymentInfo !== 'undefined' ) {
+			deprecated(
+				'Passing savePaymentInfo when registering a payment method is deprecated, instead you should pass showSavedCards and showSaveOption'
+			);
```
`deprecated` will append a deprecated since... logic into your call, so you should have it in this pattern

```
	deprecated( 'Passing a callback to registerPaymentMethod()', {
			alternative: 'a config options object',
			plugin: 'woocommerce-gutenberg-products-block',
		} );
```

Since we don't intend to remove this anytime soon, no version should be supplied.

Full docs: https://github.com/WordPress/gutenberg/tree/master/packages/deprecated

senadir reviewed (approved):
I did a code review, I will do a test review later, I really appreciate the new changed and clarity they bring.

opr committed:
Update assets/js/blocks-registry/payment-methods/payment-method-config.js

Co-authored-by: Seghir Nadir <nadir.seghir@gmail.com>

opr committed:
Add more information to deprecated call in payment method config

opr committed:
Fix lint error

nerrad commented on assets/js/base/components/payment-methods/payment-method-tab.js:
```diff
@@ -55,6 +56,7 @@ const PaymentMethodTab = ( { children, allowsSaving } ) => {
 
 PaymentMethodTab.propTypes = {
 	allowsSaving: PropTypes.bool,
+	requiresSaving: PropTypes.bool,
```
prop types need updated.

nerrad commented on assets/js/blocks-registry/payment-methods/payment-method-config.js:
```diff
@@ -21,7 +26,11 @@ export default class PaymentMethodConfig {
-			savePaymentInfo: config?.supports?.savePaymentInfo || false,
+			showSavedCards:
+				config?.supports?.showSavedCards ||
+				config?.supports?.savePaymentInfo || // Kept for backward compatibility if methods still pass this when registering.
```
++ on the comment

nerrad reviewed (approved):
Great work @opr - I agree with Nadir, this is much clearer 👏. 

I have one comment around prop types that should be addressed before merging and also we should probably add something about the `supports`  config option and the current available properties in our documentation found at `docs/extensibility/payment-method-integration.md`. 

I'll pre-approve for merging based on the code review. I have not tested, trusting in yours and Nadir's testing of things.

opr committed:
Fix prop types for PaymentMethodTab

opr committed:
Add information about supports on payment methods to docs

senadir commented:
I'm going to rebase, test, and merge this PR.
